### PR TITLE
Improve remainder traversal

### PR DESCRIPTION
**Pull request by FG-TUM**
# Description

- [x] parallelize AutoTuner::doRemainderTraversal()
  - Done via locks 
  - No-newton3 between buffers is not trivially doable because of global values
- [x] AutoPas::reserve(numParticles)
- [x] ~split LogicHandler buffers for better particle locality~
- [x] Check out task parallelization of remainder traversal.

## Comparison Before vs After

Machine: atsccs93 (i7-10700 @ 2x8x ~4.5GHz)

Config:
```yaml
verlet-rebuild-frequency         :  15
verlet-skin-radius-per-timestep  :  0.2
selector-strategy                :  Fastest-Absolute-Value
data-layout                      :  [SoA]
traversal                        :  [lc_c08]
functor                          :  Lennard-Jones (12-6) AVX intrinsics
newton3                          :  [enabled]
cutoff                           :  2.5
cell-size                        :  [1]
deltaT                           :  0.0001
iterations                       :  50
boundary-type                    :  [periodic, periodic, periodic]
Objects:                         
  CubeGrid:
    0:  
      particles-per-dimension    :  [50, 50, 50]
      particle-spacing           :  1.4
      bottomLeftCorner           :  [0.7, 0.7, 0.7]
      velocity                   :  [1000, 0, 0]
      particle-type              :  0
      particle-epsilon           :  1
      particle-sigma             :  1
      particle-mass              :  1
```

A closer look at iterations where there actually are halo particles. Average over all iterations where the particle buffer size was 2500 and halo buffer 5864. Using 8 OMP Threads.

| Timer [ms] | Before (79effeca) | After (d7e07f5) | 
|:------|-----------:|-----------------:
|Total (50 iters) | 5142 | 1585 |
| One Iteration (avg total) | 102 | 31 |
| Container::iteratePairwise (avg with halos) | ~14 | ~14 | 
| Remainder Traversal (avg with halos)| 116.8 | 4.05 |

## Related Pull Requests

- Merge after #712 

## Resolved Issues

Before this PR `doRemainderTraversal()` took a significant amount of time. If a few thousand particles were in the buffers this took significantly longer than `iteratePairwise()` due to lack of parallelization.

# How Has This Been Tested?

- [x] existing unit tests
- [x] New test for `utils::ArrayUtils::balanceVectors()`

### CODE REVIEW

File: src/autopas/utils/ArrayUtils.h [hunk 1]
```diff
@@ -146,4 +147,79 @@ std::enable_if_t<autopas::utils::ArrayUtils::is_container<Container>::value, std
...
+  // calculate vecvec statistics
+  const auto vecvecSize = vecvec.size();
+  if (vecvecSize == 0) {
+    // nothing to do
+    return;
+  }
+  const size_t numElem = std::transform_reduce(vecvec.begin(), vecvec.end(), 0, std::plus<>(),
+                                               [&](auto &vec) { return innerContainerToVec(vec).size(); });
+  const auto targetSize = static_cast<long>(numElem / vecvecSize);
+  auto rest = numElem % vecvecSize;
```
Comment (SamNewcome) [inline src/autopas/utils/ArrayUtils.h:183]: `rest` is not a clear enough name

File: src/autopas/utils/ArrayUtils.h [hunk 2]
```diff
@@ -146,4 +147,79 @@ std::enable_if_t<autopas::utils::ArrayUtils::is_container<Container>::value, std
 
   return os;
 }
+
+/**
+ * Given a collection of vectors, redistributes the elements of the vectors so they all have the same (or +1) size.
```
Comment (SamNewcome) [inline src/autopas/utils/ArrayUtils.h:152]: Add comment/note briefly describing the distribution process and make it clear that overall ordering is not preserved. E.g. Elements are taken from the ends of too-long vectors and added to the ends of too-short vectors. The overall ordering of elements is not preserved.

File: tests/testAutopas/tests/selectors/AutoTunerRemainderTraversalTest.cpp [hunk 3]
```diff
@@ -12,9 +12,298 @@
...
+ * @param particlesHaloBuffer
+ * @param n3 Newton3 on or off
+ */
+void testIteratePairwiseSteps(std::vector<Molecule> &particlesContainerOwned,
+                              std::vector<Molecule> &particlesContainerHalo,
+                              std::vector<autopas::FullParticleCell<Molecule>> &particlesBuffers,
+                              std::vector<autopas::FullParticleCell<Molecule>> &particlesHaloBuffers,
+                              autopas::Newton3Option n3) {
+  // sanity check that there are exactly two particles in the test
+  const auto numParticlesBuffers =
```
Comment (SamNewcome) [inline tests/testAutopas/tests/selectors/AutoTunerRemainderTraversalTest.cpp:33]: Rename `numParticlesBuffers` to `numParticlesInBuffers` for clarity. Similarly `numParticlesInHaloBuffers`. You _could_ also rename `particlesBuffers` to `particlesInBuffers` for consistency - but I don't think this is that important.

File: tests/testAutopas/tests/utils/ArrayUtilsTest.cpp [hunk 4]
```diff
@@ -40,3 +42,38 @@ TEST(ArrayUtilsTest, testto_string) {
...
+  const auto numElements = [&]() {
+    size_t sum = 0;
+    for (const auto &vec : vecvec) {
+      sum += vec.size();
+    }
+    return sum;
+  }();
+  // calculate expectations
+  const auto expectedPerVec = numElements / vecvec.size();
+  const auto rest = numElements % vecvec.size();
```
Comment (SamNewcome) [inline tests/testAutopas/tests/utils/ArrayUtilsTest.cpp:70]: Similar to `ArrayUtils.h`, use a different name for `rest` that's clearer

File: tests/testAutopas/tests/selectors/AutoTunerRemainderTraversalTest.cpp (hunk 3 above)
Comment (FG-TUM) [reply tests/testAutopas/tests/selectors/AutoTunerRemainderTraversalTest.cpp:33]: `particlesBuffers` are the actual buffers so I think this name makes more sense?

Comment (SamNewcome) [reply tests/testAutopas/tests/selectors/AutoTunerRemainderTraversalTest.cpp:33]: Yes I see. But I still think the `numParticlesBuffers` should be `numParticlesInBuffers` to avoid confusion that this variable does not hold the number of particle buffers.

File: src/autopas/selectors/AutoTuner.h [hunk 5]
```diff
@@ -452,120 +504,185 @@ bool AutoTuner<Particle>::iteratePairwise(PairwiseFunctor *f, bool doListRebuild
...
-                f->AoSFunctor(p1, p2, true);
-              } else {
-                // Here, we do not need to interact p1 with p2, because p is a halo particle and an AoSFunctor call with
-                // a halo particle as first argument has no effect if newton3 == false.
-                f->AoSFunctor(p2, p1, false);
-              }
+              const auto lockCoords =
+                  static_cast_array<size_t>(mulScalar(sub(p2.getR(), boxMin), interactionLengthInv));
+              // No need to apply anything to p1halo
+              //   -> AoSFunctor(p1, p2, false) not needed as it neither adds force nor Upot
```
[the remark this answers is not in the document]
Comment (FG-TUM) [reply src/autopas/selectors/AutoTuner.h]: done 6685dbbb05334e01eb1001ad4b603bcceed1b05b

File: src/autopas/utils/ArrayUtils.h (hunk 1 above)
Comment (FG-TUM) [reply src/autopas/utils/ArrayUtils.h]: done 6685dbbb05334e01eb1001ad4b603bcceed1b05b

File: src/autopas/utils/ArrayUtils.h [hunk 6]
```diff
@@ -146,4 +147,79 @@ std::enable_if_t<autopas::utils::ArrayUtils::is_container<Container>::value, std
...
+ * @param vecvec Reference to the collection of vectors to be balanced in place.
+ */
+template <class OuterContainerT>
+void balanceVectors(OuterContainerT &vecvec) {
+  balanceVectors(
+      vecvec, [](auto &innerContainer) -> auto & { return innerContainer; });
+}
+
+/**
+ * Given a collection of containers, redistributes the elements of the vectors so they all have the same (or +1) size.
```
[the remark this answers is not in the document]
Comment (FG-TUM) [reply src/autopas/utils/ArrayUtils.h]: done 6685dbbb05334e01eb1001ad4b603bcceed1b05b

File: tests/testAutopas/tests/selectors/AutoTunerRemainderTraversalTest.cpp (hunk 3 above)
Comment (FG-TUM) [reply tests/testAutopas/tests/selectors/AutoTunerRemainderTraversalTest.cpp]: done 6685dbbb05334e01eb1001ad4b603bcceed1b05b

File: tests/testAutopas/tests/utils/ArrayUtilsTest.cpp (hunk 4 above)
Comment (FG-TUM) [reply tests/testAutopas/tests/utils/ArrayUtilsTest.cpp]: done 6685dbbb05334e01eb1001ad4b603bcceed1b05b

**Review by SamNewcome — APPROVED** [review]

File: examples/md-flexible/src/TimeDiscretization.cpp [hunk 7]
```diff
@@ -44,9 +44,8 @@ void calculatePositions(autopas::AutoPas<ParticleType> &autoPasContainer,
     const auto distanceMovedSquared = dot(displacement, displacement);
     if (distanceMovedSquared > maxAllowedDistanceMoved) {
 #pragma omp critical
-      std::cerr << "A particle moved farther than skin/2/rebuildFrequency: " << std::sqrt(distanceMovedSquared) << " > "
-                << autoPasContainer.getVerletSkin() << "/2/" << autoPasContainer.getVerletRebuildFrequency() << " = "
-                << maxAllowedDistanceMoved << "\n"
+      std::cerr << "A particle moved farther than verletSkinPerTimestep/2: " << std::sqrt(distanceMovedSquared) << " > "
+                << autoPasContainer.getVerletSkinPerTimestep() << "/2 = " << maxAllowedDistanceMoved << "\n"
```
Comment (thesamriel) [inline examples/md-flexible/src/TimeDiscretization.cpp:48]: If we change the calculation here, then maybe also above in the actual calculation of `maxAllowedDistanceMoved`?
`const auto maxAllowedDistanceMoved = autoPasContainer.getVerletSkinPerTimestep() / 2`

File: examples/md-flexible/src/domainDecomposition/RegularGridDecomposition.cpp [hunk 8]
```diff
@@ -39,6 +40,16 @@ RegularGridDecomposition::RegularGridDecomposition(const MDFlexConfig &configura
 
   _decomposition = DomainTools::generateDecomposition(_subdomainCount, configuration.subdivideDimension.value);
 
+  // For reflection, we interact particles with their 'mirror' only if it would cause a repulsive force. This occurs
+  // only for particles within sixthRootOfTwo * sigma / 2.0 of the boundary. For minimal redundant iterating over
+  // particles, we iterate once over particles within the largest possible range where a particle might experience a
+  // repulsion, i.e. sixthRootOfTwo * maxSigma / 2.0.
```
Comment (thesamriel) [inline examples/md-flexible/src/domainDecomposition/RegularGridDecomposition.cpp:46]: Where do we/you know this from?

File: examples/md-flexible/src/domainDecomposition/RegularGridDecomposition.cpp [hunk 9]
```diff
@@ -388,34 +424,37 @@ RegularGridDecomposition::categorizeParticlesIntoLeftAndRightNeighbor(const std:
...
+        if (isNear(_localBoxMax[direction], _globalBoxMax[direction])) {
+          // TODO: check if this failsafe is really reasonable.
+          //  It should only trigger if a particle's position was already inside the box?
+          const auto periodicPosition = position[direction] - globalBoxLength[direction];
+          position[direction] = std::max(_globalBoxMin[direction], periodicPosition);
+          rightNeighborParticles.back().setR(position);
+        }
+      } else {
+        uncategorizedParticles.push_back(particle);
+      }
```
Comment (thesamriel) [inline examples/md-flexible/src/domainDecomposition/RegularGridDecomposition.cpp:450]: Why the change of the if-else structure?
From 
```cpp
if () ...
else if () ...
else ...
```
to 
```cpp
if () ...
else {
  if () ...
  else ...
}
```

File: src/autopas/AutoPasDecl.h [hunk 10]
```diff
@@ -130,25 +130,89 @@ class AutoPas {
...
    * @note An exception is thrown if the particle is added and it is not inside of the owned domain (defined by
    * boxmin and boxmax) of the container.
-   * @note This function is NOT thread-safe.
+   * @note This function is NOT thread-safe if the container is Octree.
    */
   void addParticle(const Particle &p);
 
+  /**
+   * Adds all particles from the collection to the container.
+   * @note This function uses addParticlesIf().
```
Comment (thesamriel) [inline src/autopas/AutoPasDecl.h:167]: I don't see where this function uses `addParticlesIf()`

Comment (thesamriel) [inline src/autopas/AutoPasDecl.h:167]: I don't see where this function uses `addParticlesIf()` 
```suggestion
   * @note This function uses addParticle().
```

File: src/autopas/AutoPasDecl.h [hunk 11]
```diff
@@ -130,25 +130,89 @@ class AutoPas {
...
    * @note An exception is thrown if the halo particle is added and it is inside of the owned domain (defined by boxmin
    * and boxmax) of the container.
-   * @note This function is NOT thread-safe.
+   * @note This function is NOT thread-safe if the container is Octree.
    */
   void addHaloParticle(const Particle &haloParticle);
 
+  /**
+   * Adds all halo particles from the collection to the container.
+   * @note This function uses addHaloParticlesIf().
```
Comment (thesamriel) [inline src/autopas/AutoPasDecl.h:197]:
```suggestion
   * @note This function uses addHaloParticle().
```

File: src/autopas/AutoPasImpl.h [hunk 12]
```diff
@@ -111,11 +111,65 @@ size_t AutoPas<Particle>::getNumberOfParticles(IteratorBehavior behavior) const
...
 }
 
+template <class Particle>
+void AutoPas<Particle>::reserve(size_t numParticles) {
+  _logicHandler->reserve(numParticles);
+}
+
+template <class Particle>
+void AutoPas<Particle>::reserve(size_t numParticles, size_t numHaloParticles) {
+  _logicHandler->reserve(numParticles);
```
Comment (thesamriel) [inline src/autopas/AutoPasImpl.h:121]:
```suggestion
  _logicHandler->reserve(numParticles, numHaloParticles);
```

File: src/autopas/CMakeLists.txt [hunk 13]
```diff
@@ -23,9 +23,10 @@ target_link_libraries(
         spdlog::spdlog
         $<$<OR:$<BOOL:${AUTOPAS_INTERNODE_TUNING}>,$<BOOL:${MD_FLEXIBLE_USE_MPI}>>:MPI::MPI_CXX>
         $<$<BOOL:${AUTOPAS_INCLUDE_MPI}>:MPI::MPI_CXX>
+        Eigen3
     PRIVATE
         # harmony and Eigen3 are only needed privately when building AutoPas.
```
Comment (thesamriel) [inline src/autopas/CMakeLists.txt:28]:
```suggestion
        # harmony is only needed privately when building AutoPas.
```

File: src/autopas/LogicHandler.h [hunk 14]
```diff
@@ -455,12 +490,12 @@ class LogicHandler {
...
    */
-  std::vector<std::vector<Particle>> _particleBuffer;
+  std::vector<FullParticleCell<Particle>> _particleBuffer;
 
   /**
    * Buffer to store halo particles that should not yet be added to the container. There is one buffer per thread.
    * @note This buffer could potentially be replaced by a ParticleCell.
    */
-  std::vector<std::vector<Particle>> _haloParticleBuffer;
+  std::vector<FullParticleCell<Particle>> _haloParticleBuffer;
```
Comment (thesamriel) [inline src/autopas/LogicHandler.h:499]: Remove note on potentially removing the buffer with ParticleCell. Or do you plan to also replace the outer vector with a ParticleCell?

Also what is the advantage of using FullParticleCell here instead of vector<>?

File: src/autopas/LogicHandler.h [hunk 15]
```diff
@@ -310,15 +346,15 @@ class LogicHandler {
...
+          additionalVectors.push_back(&(buffer._particles));
         }
       }
     }
     if (behavior & IteratorBehavior::halo) {
       for (auto &buffer : _haloParticleBuffer) {
-        if (not buffer.empty()) {
-          additionalVectors.push_back(&buffer);
+        if (not buffer.isEmpty()) {
+          additionalVectors.push_back(&(buffer._particles));
```
Comment (thesamriel) [inline src/autopas/LogicHandler.h:357]: Maybe use 
```cpp
for (auto &cell : _particleBuffer) {
auto &buffer = cell._particles;
```
 to clarify that the buffer is a cell as you did in `deleteParticlesFromBuffer()`

File: src/autopas/cells/FullParticleCell.h [hunk 16]
```diff
@@ -250,15 +248,14 @@ class FullParticleCell : public ParticleCell<Particle> {
...
    * Storage of the molecules of the cell.
    */
-  StorageType _particles;
+  StorageType _particles{};
 
   /**
    * SoA buffer of this cell.
    */
-  SoA<SoAArraysType> _particleSoABuffer;
+  SoA<SoAArraysType> _particleSoABuffer{};
```
Comment (thesamriel) [inline src/autopas/cells/FullParticleCell.h:256]: _Side note:_

Why are these members public, but use the underscore syntax, indicating a private member.  
Maybe getters would be nicer. Or is it on purpose to avoid any overhead?

File: src/autopas/utils/ParticleCellHelpers.h [hunk 17]
```diff
@@ -38,6 +38,7 @@ static bool checkParticleInCellAndUpdateByID(CellType &cell, const typename Cell
 template <class CellType>
 static bool checkParticleInCellAndUpdateByIDAndPosition(CellType &cell, const typename CellType::ParticleType &particle,
                                                         double absError) {
+  std::lock_guard<AutoPasLock> cellLock(cell._cellLock);
```
Comment (thesamriel) [inline src/autopas/utils/ParticleCellHelpers.h:41]: Do we need the lock here?
If so, I guess also in the function `checkParticleInCellAndUpdateByID()`

File: src/autopas/selectors/AutoTuner.h [hunk 18]
```diff
@@ -452,120 +510,197 @@ bool AutoTuner<Particle>::iteratePairwise(PairwiseFunctor *f, bool doListRebuild
...
 
       // 2. haloParticleBuffer with owned, close particles in container
-      for (auto &&p1 : haloParticleBuffer) {
-        auto pos = p1.getR();
-        auto min = autopas::utils::ArrayMath::subScalar(pos, cutoff);
-        auto max = autopas::utils::ArrayMath::addScalar(pos, cutoff);
+      for (auto &&p1halo : haloParticleBuffer) {
+        const auto pos = p1halo.getR();
+        const auto min = autopas::utils::ArrayMath::subScalar(pos, cutoff);
+        const auto max = autopas::utils::ArrayMath::addScalar(pos, cutoff);
```
Comment (thesamriel) [inline src/autopas/selectors/AutoTuner.h:585]:
```suggestion
        const auto min = subScalar(pos, cutoff);
        const auto max = addScalar(pos, cutoff);
```

File: src/autopas/selectors/AutoTuner.h [hunk 19]
```diff
@@ -452,120 +510,197 @@ bool AutoTuner<Particle>::iteratePairwise(PairwiseFunctor *f, bool doListRebuild
...
-            f->AoSFunctor(p1, p2, true);
+    } else {
+#ifdef AUTOPAS_OPENMP
+      // collapse loops for better scheduling
+#pragma omp for collapse(2)
+#endif
+      for (size_t i = 0; i < particleBuffers.size(); ++i) {
+        for (size_t jj = 0; jj < particleBuffers.size(); ++jj) {
+          auto *particleBufferSoAA = &particleBuffers[i]._particleSoABuffer;
+          const auto j = (i + jj) % particleBuffers.size();
```
Comment (thesamriel) [inline src/autopas/selectors/AutoTuner.h:650]: What is the reason for this? Creating diffusion for better scheduling?

File: examples/md-flexible/tests/domainDecomposition/ReflectiveBoundaryConditionTest.cpp [hunk 20]
```diff
@@ -105,31 +139,167 @@ TEST_P(ReflectiveBoundaryConditionTest, simpleReflectionTest) {
...
                                     std::array<bool, 3>{false, false, false}))
     //    ,ReflectiveBoundaryConditionTest::PrintToStringParamName());
-);
\ No newline at end of file
+);
+
+/**
+ * Implements the reflective boundary zoning test.
+ * @param particlePosition
+ * @param particleType Must be either 0 or 1. 0 corresponds to a sigma of 0.5, 1 corresponds to a sigma of 1.0.
```
Comment (thesamriel) [inline examples/md-flexible/tests/domainDecomposition/ReflectiveBoundaryConditionTest.cpp:174]:
```suggestion
 * @param particleType Must be either 0 or 1. 0 corresponds to a sigma of 0.1, 1 corresponds to a sigma of 0.2.
```

File: examples/md-flexible/tests/domainDecomposition/MixedBoundaryConditionTest.cpp [hunk 21]
```diff
@@ -47,9 +65,10 @@ auto MixedBoundaryConditionTest::setUpExpectations(
           break;
         case ::options::BoundaryTypeOption::reflective:
           // if near a reflective boundary and flying towards it the velocity sign is flipped
```
Comment (thesamriel) [inline examples/md-flexible/tests/domainDecomposition/MixedBoundaryConditionTest.cpp:67]: Comment is no longer true right?
```suggestion
          // if near a reflective boundary and flying towards it we expect a repulsive force from reflection
```

File: examples/md-flexible/src/TimeDiscretization.cpp (hunk 7 above)
Comment (FG-TUM) [reply examples/md-flexible/src/TimeDiscretization.cpp]: [d991b24](https://github.com/AutoPas/AutoPas/pull/724/commits/d991b242d7d6c4bcb195074e7cc7bc31da94f5c6)

File: examples/md-flexible/src/domainDecomposition/RegularGridDecomposition.cpp (hunk 8 above)
Comment (FG-TUM) [reply examples/md-flexible/src/domainDecomposition/RegularGridDecomposition.cpp]: Ask Sam, he wrote that :P
If you look at the LJ Potential, the distance where attraction flips to repulsion this happens at $\sqrt[6]{2} \sigma$. So for a particle to have its mirror particle in that distance, it must be $\sqrt[6]{2} \sigma / 2$ away from the mirror.

File: examples/md-flexible/src/domainDecomposition/RegularGridDecomposition.cpp [hunk 22]
```diff
@@ -252,8 +271,12 @@ void RegularGridDecomposition::exchangeMigratingParticles(AutoPasType &autoPasCo
   }
 }
 
-void RegularGridDecomposition::reflectParticlesAtBoundaries(AutoPasType &autoPasContainer) {
+void RegularGridDecomposition::reflectParticlesAtBoundaries(AutoPasType &autoPasContainer,
+                                                            ParticlePropertiesLibraryType &particlePropertiesLib) {
+  using autopas::utils::Math::isNear;
   std::array<double, _dimensionCount> reflSkinMin{}, reflSkinMax{};
+  auto functorLJ = autopas::LJFunctor<ParticleType, false, true, autopas::FunctorN3Modes::Both, false, false>(
+      _maxReflectiveSkin * 2., particlePropertiesLib);
```
[the remark this answers is not in the document]
Comment (FG-TUM) [reply examples/md-flexible/src/domainDecomposition/RegularGridDecomposition.cpp]: Since md-flexible (for now?) always does MD Simulations with the LJ potential this seems reasonable. @SamNewcome btw did we think about what the implications for the potential energy come from the boundary interactions?

File: examples/md-flexible/src/domainDecomposition/RegularGridDecomposition.cpp (hunk 9 above)
Comment (FG-TUM) [reply examples/md-flexible/src/domainDecomposition/RegularGridDecomposition.cpp]: No real reason, but also the two structures should be equivalent.
[2d81b4e](https://github.com/AutoPas/AutoPas/pull/724/commits/2d81b4eba1f70d2c841e27e450b169122a5a65bc)

File: examples/md-flexible/src/domainDecomposition/RegularGridDecomposition.cpp (hunk 22 above)
[the remark this answers is not in the document]
Comment (SamNewcome) [reply examples/md-flexible/src/domainDecomposition/RegularGridDecomposition.cpp]: I think hard coding is fine for now. Using an LJ functor for reflection becomes nasty with multi-site molecules anyway (and I suppose similarly for other rigid-body-type particles e.g. in DEM), so I was going to switch this to its own kernel anyway.

We never thought of the implications for potential energy from reflective boundaries. I would argue, though, that working around this missing potential energy is beyond the scope of md-flexible. I don't think the calculations are necessarily correct even with e.g. periodic boundaries (e.g. I don't think we add potential energy from global forces)

File: examples/md-flexible/tests/domainDecomposition/MixedBoundaryConditionTest.cpp (hunk 21 above)
Comment (SamNewcome) [reply examples/md-flexible/tests/domainDecomposition/MixedBoundaryConditionTest.cpp]: Yes. Plus the "flying towards it" part is no longer needed.

File: examples/md-flexible/src/TimeDiscretization.cpp [hunk 23]
```diff
@@ -20,9 +20,7 @@ void calculatePositions(autopas::AutoPas<ParticleType> &autoPasContainer,
   using autopas::utils::ArrayMath::dot;
   using autopas::utils::ArrayMath::mulScalar;
 
-  const auto maxAllowedDistanceMoved =
-      autoPasContainer.getVerletSkin() / (2 * autoPasContainer.getVerletRebuildFrequency());
-  const auto maxAllowedDistanceMovedSquared = maxAllowedDistanceMoved * maxAllowedDistanceMoved;
```
Comment (SamNewcome) [inline examples/md-flexible/src/TimeDiscretization.cpp:25]: Keep `maxAllowedDistanceMovedSquared`. There appears to be some error in the below code. On line 45 we compare `distanceMovedSquared` with `maxAllowedDistanceMoved` when it should be with `maxAllowedDistanceMovedSquared`

**Review by SamNewcome — CHANGES_REQUESTED** [review]

File: examples/md-flexible/tests/domainDecomposition/MixedBoundaryConditionTest.cpp (hunk 21 above)
Comment (FG-TUM) [reply examples/md-flexible/tests/domainDecomposition/MixedBoundaryConditionTest.cpp]: [2d81b4e](https://github.com/AutoPas/AutoPas/pull/724/commits/2d81b4eba1f70d2c841e27e450b169122a5a65bc)

File: examples/md-flexible/tests/domainDecomposition/ReflectiveBoundaryConditionTest.cpp (hunk 20 above)
Comment (FG-TUM) [reply examples/md-flexible/tests/domainDecomposition/ReflectiveBoundaryConditionTest.cpp]: [2d81b4e](https://github.com/AutoPas/AutoPas/pull/724/commits/2d81b4eba1f70d2c841e27e450b169122a5a65bc)

File: examples/md-flexible/src/TimeDiscretization.cpp (hunk 23 above)
Comment (FG-TUM) [reply examples/md-flexible/src/TimeDiscretization.cpp:25]: Good catch. [9024538](https://github.com/AutoPas/AutoPas/pull/724/commits/9024538819583e7b1b486e58547f288071dcc64f)

File: src/autopas/selectors/AutoTuner.h (hunk 19 above)
Comment (FG-TUM) [reply src/autopas/selectors/AutoTuner.h:650]: Yep. If we had a regular `i : 0 -> N` and `j : 0 -> N` loop all tasks that need the same `j` are created at the same time, increasing the probability of running into a lock.

File: src/autopas/cells/FullParticleCell.h (hunk 16 above)
Comment (FG-TUM) [reply src/autopas/cells/FullParticleCell.h:256]: puh... if I recall correctly this was initially a private member, hence the underscore. At some point, we made it public for ease of use. But I think we have a few naming inconsistencies like this one in the code and I think if we want to clean this up it should be a dedicated branch.

File: src/autopas/LogicHandler.h (hunk 15 above)
Comment (FG-TUM) [reply src/autopas/LogicHandler.h:357]: Those lines were only because before I changed the buffers to cells the code was
```cpp
for (auto &buffer : _particleBuffer ) {
```
Then, after the change, I needed to access the `_particles` member but since this (Cell) has the same interface as a `std::vector` adding this alias variable was a quick way to adapt the code with minimal changes ;)

Tbh I'm not sure I would change this here, because it would muddy the meaning of "buffer" even further (we'd call a member of an object in the buffer collection a buffer...)

File: src/autopas/LogicHandler.h (hunk 14 above)
Comment (FG-TUM) [reply src/autopas/LogicHandler.h]: lol yes. This is exactly what was done in this PR -> 623ba0cc4

Comment (FG-TUM) [reply src/autopas/LogicHandler.h]: The advantage is that it makes it easier to use an SoA Functor with the particles, because the cell has the logic for transforming the particle data.

File: src/autopas/CMakeLists.txt (hunk 13 above)
Comment (FG-TUM) [reply src/autopas/CMakeLists.txt]: 2fe5fe53d

**Review by thesamriel — APPROVED** [review]
